File: E_The_Lakes.cpp
#include <bits/stdc++.h>
using namespace std;

int steps[4][2] = {{0, 1}, {0, -1}, {1, 0}, {-1, 0}};

void solve() {
    int n, m, ans = 0;
    cin >> n >> m;
    vector all(n, vector<int>(m));
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            cin >> all[i][j];
        }
    }
    
    function<int(int, int)> dfs = [&](int i, int j) -> int {
        if (i < 0 || i >= n || j < 0 || j >= m || all[i][j] == 0) {
            return 0;
        }
        int sum = all[i][j];
        all[i][j] = 0;
        for (auto step : steps) {
            sum += dfs(i + step[0], j + step[1]);
        }
        return sum;
    };
    
    for (int i = 0; i < n; ++i) {
        for (int j = 0; j < m; ++j) {
            ans = max(ans, dfs(i, j));
        }
    }
    cout << ans << endl;
}

int main() {
    int T;
    cin >> T;
    while (T--) {
        solve();
    }
    return 0;
}

File: E_The_Lakes.py
dx = [0, 0, 1, -1]
dy = [1, -1, 0, 0]

for _ in range(int(input())):
    n, m = map(int, input().split())
    ans = 0
    a = [list(map(int, input().split())) for _ in range(n)]
    
    for i in range(n):
        for j in range(m):
            if a[i][j]:
                all, a[i][j] = a[i][j], 0
                q = [(i, j)]
                while q:
                    x, y = q.pop()
                    for k in range(4):
                        if 0 <= x + dx[k] < n and 0 <= y + dy[k] < m and a[x + dx[k]][y + dy[k]]:
                            all += a[x + dx[k]][y + dy[k]]
                            a[x + dx[k]][y + dy[k]] = 0
                            q.append((x + dx[k], y + dy[k]))
                ans = max(ans, all)
    print(ans)
